[Unified search] Moves dataview picker to the search bar

stratoula opened the pull request:
**This is a feature branch PR, it will merge into [elastic:unified-search-1](https://github.com/elastic/kibana/tree/unified-search-1) instead of `main`.**

Note: **We'll be tackling the full layout with this and the KQL bar in the next PR.** This is just getting the the menu combined and putting it roughly in the area of the KQL bar.

## Summary

Combines data view management actions into data view selection menu and moves data view picker for Lens and Discover to the search bar.

<img width="315" alt="Screen Shot 2022-03-15 at 17 15 31 PM" src="https://user-images.githubusercontent.com/549577/158474060-604e95ec-a169-4a10-ac89-a90f4cb066bf.png">

Adds a dismissible tour (stored in localStorage) to indicate to users that the UI has changed.

<img width="587" alt="Screen Shot 2022-03-15 at 17 15 26 PM" src="https://user-images.githubusercontent.com/549577/158474152-a6879509-b473-4f0b-962c-b768cff3f564.png">


### Notes
This PR is NOT going to be merged in main but on the `unified-search-1` feature branch. There are some additional tasks that will follow:
- Move ui folder to the data plugin --> create the `unified_search` plugin
- Move logic that is duplicated atm in both Discover and Lens to the new plugin (this can't happen at this PR due to circular dependencies)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@elasticmachine merge upstream

cchaos commented on src/plugins/discover/public/application/main/components/sidebar/discover_sidebar_responsive.tsx:
```diff
@@ -285,39 +231,11 @@ export function DiscoverSidebarResponsive(props: DiscoverSidebarResponsiveProps)
-          >
-            <EuiFlexGroup direction="row" gutterSize="s" alignItems="center" responsive={false}>
-              <EuiFlexItem grow={true}>
-                <DiscoverIndexPattern
```
So what this did was in mobile view, it added the data view selector additionally inside the flyout with all the fields. It would be really great if we could keep that functionality. Is it possible to have a second data view selector in discover in mobile view?
<img width="412" alt="Screen Shot 2022-03-10 at 18 09 59 PM" src="https://user-images.githubusercontent.com/549577/157770617-c9e96820-ff67-49d3-8114-ebb97c2cffe3.png">

stratoula replied:
Done, I am keeping it only on the flyout on the mobile view

cchaos reviewed (commented):
_Pushed you a few design updates. I'll tackle the full layout with this and the KQL bar in the other PR._

**Quick bug I found:**

When creating a new data view on the fly, after saving and closing the flyout, it's currently selected and shown in the button, but the list isn't refreshed to show it:
<img width="392" alt="Screen Shot 2022-03-10 at 17 30 11 PM" src="https://user-images.githubusercontent.com/549577/157765934-562f77e8-6c57-4d35-964d-8fc23a636826.png">

**For Lens:**

They have another instance of a data view selector in the configuration panel which is just the simple list without all the extra data view mangement stuff. But I feel like this is still a good opportunity to re-use the same data view list component. Especially since then we'll have consistent placeholder text and such. 

Is it possible at this time to make the EuiSelectable portion of the new picker standalone too?
<img width="391" alt="Screen Shot 2022-03-10 at 17 56 56 PM" src="https://user-images.githubusercontent.com/549577/157771255-3f240d95-822e-49bd-8059-edba09bdbb0f.png">

cchaos commented:
@elasticmachine merge upstream

cchaos commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@cchaos I fixed the bug and restored the datavew picker on the mobile view (Discover)
About the lens second picker is not so easy atm.
I suggest to do it when we move the ui folder from the data plugin to the `unified search` plugin. Then it will be easier to export this component and reshare it in Lens.

cchaos commented:
@gchaps In this PR we are merging the data view selector menu and the data view 'on-the-fly' menu into one. I'm also adding a dismissible tour to guide users to the new combined menu & placement which is now to the left of the KQL bar. Here are some screenshots: 

<img width="671" alt="Screen Shot 2022-03-14 at 18 18 26 PM" src="https://user-images.githubusercontent.com/549577/158270972-0acd5f41-9ff8-472a-8f28-3132f7ac8b8c.png">
<img width="382" alt="Screen Shot 2022-03-14 at 18 18 32 PM" src="https://user-images.githubusercontent.com/549577/158270988-9c7c70de-2fbb-4332-8bd3-c03181650d6f.png">



I'd love your input on copy, especially for the tour.

cchaos commented:
@stratoula I pushed a single commit that decouples just the selectable list portion into a new component for re-use in the Lens config panel. Take a look: https://github.com/elastic/kibana/pull/126560/commits/b52ce428891c0269423dab0d1f4fad731cf3dfc2 but also feel free to revert in case this is not how services work 😸

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
> * OSS CI Group #9

It seems fine Caroline :) Thanx for doing this ❤️

stratoula commented:
> @stratoula I pushed a single commit that decouples just the selectable list portion into a new component for re-use in the Lens config panel. Take a look: [b52ce42](https://github.com/elastic/kibana/commit/b52ce428891c0269423dab0d1f4fad731cf3dfc2) but also feel free to revert in case this is not how services work 😸

It seems fine Caroline :) Thanx for doing this ❤️

stratoula commented:
@elasticmachine merge upstream

gchaps commented:
t feels to me that the real value add is that the two menus were combined in a single menu and that the trigger is on the menu that hasn't moved.  Also, I feel there needs to be a more descriptive title than "Hey there".  What about something like either of these:

**A better data view menu**
Find, edit, manage, and create data views from a single menu.

**A better data view menu**
All options for working with your data views are now in one place.

gchaps commented on src/plugins/data/public/ui/dataview_picker/change_dataview.tsx:
```diff
@@ -0,0 +1,221 @@
+                    }}
+                  >
+                    {i18n.translate('data.query.queryBar.indexPattern.addFieldButton', {
+                      defaultMessage: 'Add a field to this data view...',
```
We don't typically  include ... after menu items

```suggestion
                      defaultMessage: 'Add a field to this data view',
```

cchaos replied:
Hey @gchaps , so about this. What I'm trying to establish here is a pattern of indicating when menu items will open up a modal/flyout. I've seen this pattern in other places like Google docs that are heavy on menus. It basically indicates that the menu item will not **perform** the action but bring up a longer dialog. I know it's a bit confusing since we also truncate a lot with ellipses. Do you know of any other pattern we can start utilizing?

gchaps replied:
My thought is that using the right verb is enough to suggest what is going to happen from the menu (like opening a popup or a wizard or something outside of the menu itself).  Seeing a menu with a lot of items followed by ... makes the menu harder to read and scan.

The ellipses is the best pattern to indicate when a menu items opens a modal/flyout. It's commonly used by other companies like Microsoft, Apple, and Google to indicate that more input is required. 

We haven't used the ellipsis on menu items in the rest of the UI.  If we start using them here, I think we should review all other menus as well. For example, the Options menu on a dashboard panel has two items that open a flyout or a modal.

cchaos replied:
> We haven't used the ellipsis on menu items in the rest of the UI. If we start using them here, I think we should review all other menus as well.

Yep, exactly. I would like to start a recognizable pattern.

> using the right verb is enough to suggest what is going to happen from the menu

While I disagree generally with this statement, I don't have a strong enough reasoning in this particular menu. But I'd definitely like to start a general discussion around menu items. I'll think about starting a doc, but for now I"ll remove them.

gchaps replied:
+1 to starting a discussion around menu items and making them consistent.

gchaps commented on src/plugins/data/public/ui/dataview_picker/change_dataview.tsx:
```diff
@@ -0,0 +1,221 @@
+                    }}
+                  >
+                    {i18n.translate('data.query.queryBar.indexPattern.manageFieldButton', {
+                      defaultMessage: 'Manage this data view...',
```
```suggestion
                      defaultMessage: 'Manage this data view',
```

gchaps commented on src/plugins/data/public/ui/dataview_picker/change_dataview.tsx:
```diff
@@ -0,0 +1,221 @@
+                }}
+              >
+                {i18n.translate('data.query.queryBar.indexPattern.addNewDataView', {
+                  defaultMessage: 'New data view...',
```
```suggestion
                  defaultMessage: 'New data view',
```

gchaps replied:
Should this be "Create a data view"?

cchaos replied:
Yep, sorry bad copy from mock, should have stuck with existing text.

gchaps commented:
@cchaos For the copy for the popup, which of the three is your favorite option?

cchaos commented:
Thanks @gchaps! I've updated the screenshots in PR summary. I went with your last option for the tour text. Are you also good with the dismiss text being `Got it!`?

gchaps commented:
`Got it` is good, but I don't think it needs the exclamation point.

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@elasticmachine merge upstream

kibanamachine commented:
expected head sha didn’t match current head ref.

stratoula commented:
@elastic/kibana-core I am trying to figure out why the Jest integration tests are failing [here](https://buildkite.com/elastic/kibana-pull-request/builds/31345#8738bea7-f31a-4428-b7b3-67e6a1f4ebf2) 

specifically this test `src/core/server/saved_objects/migrations/actions/integration_tests/actions.test.ts`. Do you have any idea? 🤔

stratoula commented:
@elasticmachine merge upstream

kibanamachine commented:
expected head sha didn’t match current head ref.

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
@elasticmachine merge upstream

kertal reviewed (commented):
Great work @stratoula, added a minor comment about the code, and some findings when testing it (since this will be merged into a feature branch, these are no blockers)

So when testing in Discover using Safari, here are my thoughts, high likely that you are already aware of them:
<img width="602" alt="Discover_-_Elastic" src="https://user-images.githubusercontent.com/463851/158981113-d1698842-ee90-406d-b4c6-debb581809dd.png">
<img width="917" alt="Discover_-_Elastic_und__Unified_search__Moves_dataview_picker_to_the_search_bar_by_stratoula_·_Pull_Request__126560_·_elastic_kibana" src="https://user-images.githubusercontent.com/463851/158981615-555a2a59-dca6-4c95-9d6f-2bfccc275941.png">

stratoula commented on src/plugins/discover/public/application/main/components/sidebar/discover_sidebar.tsx:
```diff
@@ -282,34 +282,6 @@ export function DiscoverSidebarComponent({
     return null;
   }
 
-  if (useFlyout) {
```
Correct and done!

stratoula commented:
@matthias thanx a lot for the review about the
![image](https://user-images.githubusercontent.com/17003240/158987173-c789cf18-50d8-437d-b5f6-f3e1f9deb9a0.png)

dont worry about it because we have a follow up PR https://github.com/elastic/kibana/pull/126566 that makes the timepicker to have an auto width and I think it will fix this problem. Let's revisit this then.
![image](https://user-images.githubusercontent.com/17003240/158987339-79614949-8c26-4815-890f-2a18204a3653.png)

About the mobile view your comment makes sense.
@cchaos wdyt about it?

![image](https://user-images.githubusercontent.com/17003240/158987447-1817180b-d19f-4016-a821-b4a1624cc4f0.png)

stratoula commented:
I actually implemented @kertal's suggestions for mobile view. It looks like that now
<img width="568" alt="image" src="https://user-images.githubusercontent.com/17003240/158991953-1a3b1fef-5b1c-4314-943f-d053c0e64162.png">

kertal commented:
> I actually implemented @kertal's suggestions for mobile view. It looks like that now
> <img alt="image" width="568" src="https://user-images.githubusercontent.com/17003240/158991953-1a3b1fef-5b1c-4314-943f-d053c0e64162.png">

Thx @stratoula, for me, it looks more logical this way (Nit: bottom margin of data view could be similar to the time picker bottom margin)

Interesting, because it never drew my attention until now:

In the `normal` view, the order is 

`query bar `| `time filter`

But in the mobile view 

`time filter` |` query bar `

I guess this switch was because of filters would look ugly in a sandwich of query bar and time filter, and I'm pretty sure when unified search progresses this order won't be different in `normal` and `mobile` view

stratoula commented:
Exactly, I also noticed that now :) This is the reason I decided to proceed with your suggestion and not leave it for later :D
Not the order is respected. About the margin I dont have a strong opinion. I will wait for Caroline's feedback on the changes!

cchaos commented:
Thanks @kertal , we'll be taking care of the final layout in the next PR. I thought I had put in a sentence in the summary regarding that, but I guess not... sorry about the confusion.

kertal commented:
> Thanks @kertal , we'll be taking care of the final layout in the next PR. I thought I had put in a sentence in the summary regarding that, but I guess not... sorry about the confusion.

Oh, no worries, no confusion, just wanted to express my first impressions, knowing this is not the final stage, no blocker in sight from data discovery. great work!

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

stratoula commented:
@elasticmachine merge upstream

flash1293 commented on x-pack/plugins/lens/public/app_plugin/lens_top_nav.tsx:
```diff
@@ -548,6 +573,122 @@ export const LensTopNavMenu = ({
+    }
+    closeDataViewEditor.current = dataViewEditor.openEditor({
+      onSave: async (dataView) => {
+        if (dataView.id) {
```
nit: double check?

flash1293 commented on x-pack/plugins/lens/kibana.json:
```diff
@@ -20,6 +20,7 @@
     "share",
     "presentationUtil",
     "dataViewFieldEditor",
+    "dataViewEditor",
```
It seems like this is only used for the permissions, is there any difference between `userPermissions` of `dataViewEditor` and `dataViewFieldEditor`? The implementation looks identical

stratoula replied:
It is also used for opening the dataView editor which is not the same. From my understanding these are two different editors with different permissions.

flash1293 replied:
A right, makes sense

flash1293 reviewed (approved):
I want to echo the thing @kertal mentioned - index patterns can get quite long and we should restrict the size somehow:
<img width="2545" alt="Screenshot 2022-03-21 at 11 22 29" src="https://user-images.githubusercontent.com/1508364/159242757-cb58512e-4224-4326-848d-9a169c506e85.png">

On creating a new data view from within Lens (yay for adding that as a side effect), it doesn't tell the app to reload its internal data view list:
I created the super long one from within the flyout, but it's not showing in the layer dropdown:
<img width="1541" alt="Screenshot 2022-03-21 at 11 23 50" src="https://user-images.githubusercontent.com/1508364/159242923-4a1e765d-3ceb-4598-87bf-cc6686c65f2d.png">

Might be unrelated but the "create data view" flyout is extremely laggy for me.

No blockers from my side as this is going to a feature branch but we should make sure to address before the final merge

stratoula commented:
Update from @flash1293 's comments:
- I reverted the picker full width to the max-width option. Let's check the mobile view on the next PR https://github.com/elastic/kibana/pull/126566
- I fixed the bug with the layerpanel dataview list not being updated when a new dataview is created
- Removed the double dataview.id check
